src/HealthChecks.AzureServiceBus/AzureEventHubHealthCheck.cs

```diff
@@ -1,63 +1,73 @@
+    private readonly string? _endpoint;
+    private readonly string? _eventHubName;
+    private readonly TokenCredential? _tokenCredential;
+    private readonly EventHubConnection? _connection;
+
+    private string? _connectionKey;
+
+    private string ConnectionKey =>
+        _connectionKey ??= _connectionString ?? $"{_endpoint}_{_eventHubName}";
+
+    public AzureEventHubHealthCheck(string connectionString, string eventHubName)
     {
-        private const string ENTITY_PATH_SEGMENT = "EntityPath=";
+        Guard.ThrowIfNull(connectionString, true);
+        Guard.ThrowIfNull(eventHubName, true);
 
-        private static readonly ConcurrentDictionary<string, EventHubProducerClient> _eventHubConnections = new ConcurrentDictionary<string, EventHubProducerClient>();
-        private readonly string _eventHubConnectionString;
+        _connectionString = connectionString.Contains(ENTITY_PATH_SEGMENT) ? connectionString : $"{connectionString};{ENTITY_PATH_SEGMENT}{eventHubName}";
+    }
 
-        public AzureEventHubHealthCheck(string connectionString, string eventHubName)
-        {
-            Guard.ThrowIfNull(connectionString, true);
-            Guard.ThrowIfNull(eventHubName, true);
+    public AzureEventHubHealthCheck(EventHubConnection connection)
+    {
+        Guard.ThrowIfNull(connection);
 
-            _eventHubConnectionString = connectionString.Contains(ENTITY_PATH_SEGMENT) ? connectionString : $"{connectionString};{ENTITY_PATH_SEGMENT}{eventHubName}";
+        _connectionKey = $"{connection.FullyQualifiedNamespace}_{connection.EventHubName}";
+        _connection = connection;
+    }
 
-            if (!_eventHubConnections.ContainsKey(_eventHubConnectionString))
-            {
-                _eventHubConnections.TryAdd(_eventHubConnectionString, new EventHubProducerClient(_eventHubConnectionString));
-            }
-        }
+    public AzureEventHubHealthCheck(string endpoint, string eventHubName, TokenCredential tokenCredential)
+    {
+        _endpoint = Guard.ThrowIfNull(endpoint, true);
+        _eventHubName = Guard.ThrowIfNull(eventHubName, true);
+        _tokenCredential = Guard.ThrowIfNull(tokenCredential);
+    }
 
-        public AzureEventHubHealthCheck(EventHubConnection connection)
+    public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+    {
+        try
         {
-            Guard.ThrowIfNull(connection);
-
-            _eventHubConnectionString = $"{connection.FullyQualifiedNamespace};{ENTITY_PATH_SEGMENT}{connection.EventHubName}";
-
-            if (!_eventHubConnections.ContainsKey(_eventHubConnectionString))
-            {
-                _eventHubConnections.TryAdd(_eventHubConnectionString, new EventHubProducerClient(connection));
-            }
-        }
+            var client = await ClientCache.GetOrAddAsyncDisposableAsync(ConnectionKey,
+                async _ => await Task.FromResult(CreateClient()));
```

Comment:
I would not juggle async/await here. Just add another ` public static async ValueTask<T> GetOrAddAsyncDisposableAsync<T>(string key, Func<string, T> clientFactory) where T : IAsyncDisposable` overload and use it here. `IAsyncDisposable` is about asyncronous destroying, not about asyncronous creation.

Reply:
That sounds more sane 😄 Done.